include/flash_pm.h

```diff
@@ -0,0 +1,46 @@
+/*
```

Comment:
Actually I intended to ifdef this file so that when you select alternative PM from Kconfig, FLASH_AREA macros from this file would not be included, instead some user writable header file would be included, similar way the zephyr/include/generated/ files are included during build. This way PM could provide it's own generated file that would contain definitions and FLASH_AREA macros to support them in code. I can do that from flash_map.h

Reply:
>  when you select alternative PM from Kconfig, FLASH_AREA macros from this file would not be included

I don't think that's going to fly, since standard practice now is to define flash areas in devicetree.  If you want a way to *extend* the available flash areas through some sort of partition manager that should be fine, but eliminating the ability to simultaneously specify others in devicetree might be a hard sell.

I suspect it'd be no more than putting:

    #ifdef CONFIG_EXTERNAL_PM
    #include <path/to/generated/pm.h>
    #endif

within the `enum flash_area_partitions {` declaration.

Reply:
I disagree @pabigot , IMO you pick one or the other. Today we overwrite the partitions defined in device tree when enabling partition manager. This is the only way we get full control of the layout of the device.

Reply:
> This is the only way we get full control of the layout of the device.

I'd need to see much more detail on exactly what the proposed partition manager for upstream Zephyr would look like.  Pending that detail:
* If you want to have a partition manager have full control of the layout of a given device, I can see that making sense.
* Association of a partition manager with one device must not prevent use of immutable device-tree--defined region definitions (as we have today) on another device.

To make an analogy with Linux: It's great to be able to put LVM on a block device so I can split it up for different needs that change over time, but we still need fixed partition tables (MSDOS or GPT), and to be able to put a filesystem on the entire device without a partition table at all.

On initial consideration I would think the flexible way is to continue to define regions in device tree, and specify that a partition manager can further subdivide regions that are assigned to it (which may be the span of an entire device).

This leads to a question of why flash areas aren't simply treated as devices in their own right, delegating to a parent device, but that's something for the future, possibly coordinating with RAM regions.

None of that appears immediately relevant to this PR which AIUI adds a layer of abstraction for retrieving information about a flash region's definition, starting with support for the name/id-based devicetree information we're using today.  But please keep it in mind for whatever follow-on work is envisioned.

Reply:
> * Association of a partition manager with one device must not prevent use of immutable device-tree--defined region definitions (as we have today) on another device.

Are you talking about multiple devices in the same build (ie cmake invocation) or in separate builds? In a single build it is still possible to "opt out" of partition manager (simply don't enable the USE_PARTITION_MANAGER symbol). If you are talking about another device in another built context, then there is no issue as we decide whether or not to invoke something like partition manager on a per build basis.

> To make an analogy with Linux: It's great to be able to put LVM on a block device so I can split it up for different needs that change over time, but we still need fixed partition tables (MSDOS or GPT), and to be able to put a filesystem on the entire device without a partition table at all.

Note that none of these configurations will change in run time. It can be modified during a firmware update, but I don't consider than runtime. The filesystems (or LVM in this case)
 
> On initial consideration I would think the flexible way is to continue to define regions in device tree, and specify that a partition manager can further subdivide regions that are assigned to it (which may be the span of an entire device).

This is not sufficient as we, as stated above, need full control of the partitions on the device.
DT and PM solve the same part of partitioning, only in different ways.

> This leads to a question of why flash areas aren't simply treated as devices in their own right, delegating to a parent device, but that's something for the future, possibly coordinating with RAM regions.

Are you thinking about a runtime configuration like GPT?

Reply:
I'm saying that if PM is used on the SOC address space, I should still be able to use DT to "partition" an external SPI flash that is a completely different device.

There is nothing in this PR that seems to conflict with that, and I'm fine with this PR once the macros move to `flash_map.h` where the functions that use flash area material generally live.